### PR TITLE
[JIT] Improve BatchMM mutability handling

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64955

Previously, BatchMM would skip any block containing any mutable
operators. Now it will avoid batching any operation whose inputs or
outputs are ever mutated.

Tests added:
```
python test/test_jit.py TestBatchMM
```

Differential Revision: [D30913266](https://our.internmc.facebook.com/intern/diff/D30913266)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30912798).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30913266).

File: torch/csrc/jit/passes/batch_mm.cpp [hunk 1]
```diff
@@ -432,7 +437,8 @@ void BatchMMSide(Block* block, AliasDb& alias_db) {
 
   std::unordered_set<Value*> considered_values;
   for (Node* node : block->nodes()) {
-    if (node->matches("aten::mm(Tensor self, Tensor mat2) -> Tensor")) {
+    if (node->matches("aten::mm(Tensor self, Tensor mat2) -> Tensor") &&
+        !alias_db.hasWriters(node)) {
       for (Value* input : node->inputs()) {
         if (/*bool not_inserted = */ !considered_values.emplace(input).second) {
           continue;
```
Comment (eellison) [inline torch/csrc/jit/passes/batch_mm.cpp:444]: Do we need to add a check in `removeTransposesAndGatherMatmuls` ?

**Review by eellison — COMMENTED** [review]
Looks great !! a bunch of minor comments and one question about why we're not generating a partial batching in one of the tests

File: torch/csrc/jit/passes/batch_mm.cpp [hunk 2]
```diff
@@ -465,13 +471,26 @@ bool hasMutableOperators(Block* block) {
   return false;
 }
 
+bool hasMMOperators(Block* block) {
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/csrc/jit/passes/batch_mm.cpp]: Cool, thanks for point this out - I've changed it

File: test/jit/test_batch_mm.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                torch.mm(T1, T2)
+                + torch.mm(T3, T4)
+                + torch.mm(T5, T6)
+                + torch.mm(T7, T8)
+            )
+            return result
+
+        FileCheck().check_count("aten::mm", 4, False).run(str(test_batch_mm.graph))
+        self.run_pass("batch_mm", test_batch_mm.graph)
+        FileCheck().check_count("aten::mm", 4, False).run(str(test_batch_mm.graph))
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply test/jit/test_batch_mm.py]: Whoops, fixed this.

File: test/jit/test_batch_mm.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            T3 = torch.zeros((n, n))
+            T4 = torch.zeros((n, n))
+            T5 = torch.zeros((n, n))
+            T6 = torch.zeros((n, n))
+            T7 = torch.zeros((n, n))
+            T8 = torch.zeros((n, n))
+            torch.relu_(T1)
+            result = (
+                torch.mm(T1, T2)
+                + torch.mm(T3, T4)
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply test/jit/test_batch_mm.py:79]: The logic I was aiming to implement is as follows: "Consider the tree of MM, T, and ADD nodes rooted at some ADD node. If any input or output value to any of the nodes is mutated in the graph, then ignore the whole tree".  This approach is definitely way too general, but I also think that "checking that none of the inputs to a matrix multiplication are written to" is probably not sufficient: for example, consider the following:
```
%1 = MM(T1, T2)
%2 = MM(T3, T4)
%3 = ADD(%1, %2)
%4 = MM(T5, T6)
%5 = ADD(%3, %4)
RELU_(%5)
%6 = MM(T7, T8)
%7 = ADD(%5, %6)
```

Perhaps a better approach is "in a tree of MM, T, and ADD nodes rooted at some ADD node, ignore any subtree rooted at a node whose inputs are modified" - does this sound correct?

File: test/jit/test_batch_mm.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                + torch.mm(T5, T6)
+                + torch.mm(T7, T8)
+            )
+            result["constant"] = torch.tensor([42.0])
+            return result
+
+        print(str(test_batch_mm))
+
+        FileCheck().check_count("aten::mm", 4, True).run(str(test_batch_mm.graph))
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply test/jit/test_batch_mm.py]: Thanks - I've added this for this test and the previous one.

Another question about this - do you have a suggestion on a good way to to pass in or generate tensors `T1..T8` used in the test? When I provide them as parameters, I notice that `AliasDb` considers a mutation to any single parameter to be a mutation of all the parameters. I can do something messy (like including literal tensor definitions or constructing them based on input parameters) - but can you suggest a better way?  It's not an issue in this test but in some of the other tests it matters.

File: test/jit/test_batch_mm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply test/jit/test_batch_mm.py:109]: Cool, yea good point. I checked this test case without the `torch.mm(T1, T2)` included in the result and batchmm doesn't fire, so I guess there's other criteria that is not being hit if we just exclude it. 

>  a tree of MM, T, and ADD nodes rooted at some ADD node, ignore any subtree rooted at a node whose inputs are modified

That sounds right to me... however the PR as it stands is already a big jump in handling mutability, and I'm not convinced that we need to go too down the rabbit hole with maximally handling mutation. I think it is fine to land as is.

File: torch/csrc/jit/passes/batch_mm.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/batch_mm.cpp:445]: I guess not because this is just examining a tree we've already verified to not have writers

**Review by eellison — APPROVED** [review]
Looks great!! 🚢 🚢 🚢

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30913266).

**Pull request merged**